webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/users/UserRepository.scala

```diff
@@ -0,0 +1,834 @@
+
+  implicit val userWrite: Write[UserInfo] = {
+    Write[
+      (String, DateTime, String, String, Option[String], Option[String], Option[DateTime], List[StatusHistory], Json.Obj)
+    ].contramap {
+      case u =>
+        (u.id, u.creationDate, u.status.value, u.managedBy, u.name, u.email, u.lastLogin, u.statusHistory, u.otherInfo)
+    }
+  }
+
+  implicit val userRead: Read[UserInfo] = {
+    Read[
+      (String, DateTime, String, String, Option[String], Option[String], Option[DateTime], List[StatusHistory], Json.Obj)
+    ].map {
+      u: ((String, DateTime, String, String, Option[String], Option[String], Option[DateTime], List[StatusHistory], Json.Obj)) =>
+        UserInfo(
+          u._1,
+          u._2,
+          UserStatus.parse(u._3).toOption.getOrElse(UserStatus.Disabled),
+          u._4,
+          u._5,
+          u._6,
+          u._7,
+          u._8,
+          u._9
+        )
+    }
+  }
+
+  // when we create a session for an user, we also update his "lastLogin" tom
+  override def logStartSession(
+      userId:            String,
+      permissions:       List[String],
+      sessionId:         SessionId,
+      authenticatorName: String,
+      date:              DateTime
+  ): IOResult[Unit] = {
+    val session   = {
+      sql"""insert into usersessions (sessionid, userid, creationdate, authmethod, permissions)
+            values (${sessionId}, ${userId}, ${date}, ${authenticatorName}, ${permissions.sorted})"""
+    }
+    val lastLogin = {
+      sql"""update users set lastlogin = ${date} where id = ${userId}"""
+    }
+
+    transactIOResult(s"Error when saving session '${sessionId.value}' info for user '${userId}'")(xa => {
+      (for {
+        _ <- session.update.run
+        _ <- lastLogin.update.run
+      } yield ()).transact(xa)
+    }).unit
+  }
+
+  override def logCloseSession(userId: String, endDate: DateTime, endCause: String): IOResult[Unit] = {
+    val sql =
+      sql"""update usersessions set enddate = ${endDate}, endcause = ${endCause} where enddate is null and userid = ${userId}"""
+    transactIOResult(s"Error when closing opened session for user '${userId}'")(xa => sql.update.run.transact(xa)).unit
+  }
+
+  override def closeAllOpenSession(endDate: DateTime, endCause: String): IOResult[Unit] = {
+    val sql = sql"""update usersessions set enddate = ${endDate}, endcause = ${endCause} where enddate is null returning userid"""
+    transactIOResult(s"Error when closing opened user session")(xa => sql.query[String].to[List].transact(xa)).flatMap(users =>
+      ApplicationLoggerPure.User.info(s"Close open sessions with reason '${endCause}' for users '${users.mkString("', '")}'")
+    )
+  }
+
+  override def getLastPreviousLogin(userId: String): IOResult[Option[UserSession]] = {
+    val sql =
+      sql"""select * from usersessions where userid = ${userId} and enddate is not null order by creationdate desc limit 1"""
```

Comment:
The "*" is not a good practice and now we know why : the `tenants` column should not be included in 7.3 but only since 8.1...

```suggestion
      sql"""select userid, sessionid, creationdate, authmethod, permissions, enddate, endcause from usersessions where userid = ${userId} and enddate is not null order by creationdate desc limit 1"""
```


One more reliable and maintainable way to do `select *` in multiple queries is to declare it as a static fragment value : 
```scala
// maps to `UserSession` case class
val all: Fragment = fr"userid, sessionid, creationdate, authmethod, permissions, enddate, endcause"
```
, which would be in 8.1 : 
```scala
// maps to `UserSession` case class
val all: Fragment = fr"userid, sessionid, creationdate, authmethod, permissions, tenants, enddate, endcause"
```